docs/cookbook/squad_extraction.md

```diff
@@ -0,0 +1,187 @@
+
+The first step is loading the dataset. We've written a helper function `load_geology_squad` that will load and return a `list[QuestionWithContext]`, defined below:
+
+```python
+from pydantic import BaseModel
+
+class QuestionWithContext(BaseModel):
+    """A question with it's answers and context."""
+
+    id: str
+    question: str
+    context: str
+    answers: list[str]
+```
+
+## Initial Basic Extraction
+
+To extract an answer to a question, we can use the [`OpenAIChat.extract`](../api/chat/models/openai_chat.md/#mirascope.chat.models.OpenAIChat.extract) method to extract the answer. First, let's create a super basic schema and prompt for asking the question and extracting an answer:
+
+```shell
+$ mirascope init; touch prompts/question.py
+```
+
+```python
+# prompts/question.py
+from mirascope import Prompt
+from pydantic import BaseModel, Field
+
+
+class ExtractedAnswer(BaseModel):
+    """The anwer to a question about a paragraph of text."""
+
+    answer: str
+
+
+class QuestionPrompt(Prompt):
+    """
+    Paragraph: {paragraph}
+
+    Question: {question}
+    """
+
+    paragraph: str
+    question: str
+```
+
+Next we'll define the schema we want to extract from the paragraph and a function to extract the schema from a given question:
+
+```python
+from pydantic import BaseModel
+
+from config import Settings
+from prompts.question import ExtractedAnswer, QuestionPrompt
+
+settings = Settings()
+chat = OpenAIChat(model="gpt-3.5-turbo-1106", api_key=settings.openai_api_key)
+
+
+def extract_answer(question: QuestionWithContext) -> str:
+    """Returns the extracted `str` answer to `question`."""
+    return chat.extract(
+        ExtractedAnswer,
+        QuestionPrompt(paragraph=question.context, question=question.question),
+        retries=2,  # retry up to 2 more times on validation error
+    ).answer
+```
+
+Now we just need to load the data and extract our answers:
+
+```python
+import json
+
+from squad import load_geology_squad
+
+extracted_answers = {
+    question.id: extract_answer(question)
+    for question in load_geology_squad()
+}
+with open("geology-squad-answers-v1.json", "w") as f:
+    json.dump(extracted_answers, f)
+```
+
+Running the included modified version of the SQuAD eval script gives us the following breakdown on metrics, which we can use to see how well our system is working:
+
+```shell
+$ python eval.py geology-squad.json geology-squad-answers-v1.json
+```
+
+```
+{
+  "exact": 79.3103448275862,
+  "f1": 89.4777077966733,
+  "total": 116,
+}
+```
+
+Exact match is determined with a raw string match, whereas F1 score is calculated by token to account for partial matches. With our simple script, we were able to get nearly 80% exact match and an F1 score of nearly 90 on our 116 examples. Before we move on, let's include this metadata in the docstring of our prompt file and version it with the CLI:
```

Comment:
90 -> missing % sign?

Reply:
This is F1 score, so no percent (vs. exact match which is a percentage).